lektor/admin/static/js/widgets/tuiEditorLektorImageExt.jsx

```diff
@@ -0,0 +1,307 @@
+/*
+ * Insert Image extension for TUI editor
+ * Based on popupAddImage.js and colorSyntax.js
+ *
+ * Integrates with Lektor's attachment system to enable selecting an existing
+ * attachment, and uploading a new image attachment.
+ */
+
+import React from 'react'
+
+import hub from '../hub'
+import utils from '../utils'
+import { AttachmentsChangedEvent } from '../events'
+import RecordComponent from '../components/RecordComponent'
+import makeRichPromise from '../richPromise'
+
+// name of the extension
+const EXT_NAME = 'lektorLink'
+
+class LektorImageExtension extends RecordComponent {
+  constructor (props) {
+    super(props)
+
+    this.editor = null
+    this.button = null
+
+    this.initForm = {
+      attachmentPath: '----',
+      attachmentAltText: '',
+      uploadFile: null,
+      uploadAltText: '',
+      externalPath: '',
+      externalAltText: ''
+    }
+
+    this.state = {
+      show: false,
+      top: '',
+      left: '',
+      mode: 'attachment',
+      files: [],
+      form: this.initForm
+    }
+
+    this.modeChange = this.modeChange.bind(this)
+    this.formChange = this.formChange.bind(this)
+    this.canAddImage = this.canAddImage.bind(this)
+    this.addImage = this.addImage.bind(this)
+    this.init = this.init.bind(this)
+    this.hide = this.hide.bind(this)
+    this.show = this.show.bind(this)
+  }
+
+  // called when we switch tabs
+  modeChange (e) {
+    this.setState({
+      mode: e.target.getAttribute('data-mode')
+    })
+  }
+
+  // called whenever one of the controlled form elements is updated
+  formChange (e) {
```

Comment:
In `ToastEditor` I think this is called `onChange`. Is there a reason they are different?

Reply:
No reason, and I've changed it to `onChange` to be more consistent with both lektor and React convention.